File: ios-app/Packages/BayitCast/Sources/BayitCast/Integration/MediaPlayerCastBridge.swift
import BayitCore
import BayitMedia
import Combine
import Foundation

/// Bridges MediaPlayer with CastSessionManager to sync playback state.
@MainActor
public final class MediaPlayerCastBridge: ObservableObject {

    private let mediaPlayer: MediaPlayer
    private let castManager: CastSessionManager
    private let logger = BayitLogger(category: "MediaPlayerCastBridge")

    private var cancellables = Set<AnyCancellable>()
    private var lastSyncTime: TimeInterval = 0
    private let syncThreshold: TimeInterval = 1.0

    @Published public private(set) var isCasting: Bool = false

    public init(
        mediaPlayer: MediaPlayer,
        castManager: CastSessionManager
    ) {
        self.mediaPlayer = mediaPlayer
        self.castManager = castManager
        setupObservers()
    }

    private func setupObservers() {
        castManager.statePublisher
            .sink { [weak self] state in
                self?.handleCastStateChange(state)
            }
            .store(in: &cancellables)

        Timer.publish(every: 5.0, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                Task { @MainActor in
                    await self?.syncPlaybackState()
                }
            }
            .store(in: &cancellables)
    }

    private func handleCastStateChange(_ state: CastSessionState) {
        isCasting = state.isConnected

        logger.info("Cast state changed", context: [
            "newState": state.rawValue,
            "isCasting": "\(isCasting)"
        ])

        if state.isConnected {
            Task {
                await loadCurrentMedia()
            }
        }
    }

    private func loadCurrentMedia() async {
        guard let currentItem = mediaPlayer.avPlayer.currentItem,
              let url = (currentItem.asset as? AVURLAsset)?.url else {
            logger.warning("Cannot load media to cast - no current item")
            return
        }

        let media = CastMedia(
            contentId: UUID().uuidString,
            title: "Bayit+ Content",
            streamUrl: url,
            posterUrl: nil,
            duration: mediaPlayer.duration > 0 ? mediaPlayer.duration : nil
        )

        do {
            try await castManager.loadMedia(media)
            await syncPlaybackState()
            logger.info("Current media loaded to cast device")
        } catch {
            logger.error("Failed to load current media to cast device", error: error)
        }
    }

    public func syncPlaybackState() async {
        guard isCasting else { return }

        let currentTime = mediaPlayer.currentTime
        let timeDiff = abs(currentTime - lastSyncTime)

        guard timeDiff >= syncThreshold else {
            return
        }

        let playbackState = CastPlaybackState(
            currentTime: currentTime,
            isPlaying: mediaPlayer.state == .playing,
            volume: 1.0
        )

        do {
            try await castManager.syncPlaybackState(playbackState)
            lastSyncTime = currentTime
        } catch {
            logger.error("Failed to sync playback state to cast", error: error)
        }
    }

    public func loadMedia(_ media: CastMedia) async throws {
        guard isCasting else {
            throw CastError.notConnected
        }

        try await castManager.loadMedia(media)
        logger.info("Media loaded to cast device via bridge", context: [
            "contentId": media.contentId
        ])
    }
}

File: android-app/core/core-cast/src/main/java/tv/bayit/plus/core/cast/MediaPlayerCastBridge.kt
package tv.bayit.plus.core.cast

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import tv.bayit.plus.core.cast.models.CastMedia
import tv.bayit.plus.core.cast.models.CastPlaybackState
import tv.bayit.plus.core.cast.models.CastSessionState
import tv.bayit.plus.core.cast.models.CastSubtitleTrack
import tv.bayit.plus.core.common.logging.BayitLogger
import tv.bayit.plus.core.media.BayitMediaPlayer
import tv.bayit.plus.core.media.PlayerState
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MediaPlayerCastBridge @Inject constructor(
    private val castSessionManager: CastSessionManager,
    private val logger: BayitLogger,
) {
    private var syncJob: Job? = null
    private var observeJob: Job? = null
    private var lastSyncTimeMs: Long = 0L
    private var attachedPlayer: BayitMediaPlayer? = null
    private var currentContentId: String? = null
    private var currentTitle: String? = null

    fun isCastingNow(): Boolean = castSessionManager.sessionState.value.isConnected

    fun attach(player: BayitMediaPlayer, scope: CoroutineScope) {
        detach()
        attachedPlayer = player
        observeJob = scope.launch {
            castSessionManager.sessionState.collect { state ->
                handleStateChange(state, scope)
            }
        }
    }

    fun detach() {
        syncJob?.cancel()
        observeJob?.cancel()
        syncJob = null
        observeJob = null
        attachedPlayer = null
    }

    fun updateContent(contentId: String, title: String) {
        currentContentId = contentId
        currentTitle = title
    }

    fun updateContent(
        contentId: String,
        title: String,
        streamUrl: String,
        contentType: String,
        subtitleTracks: List<CastSubtitleTrack>,
        scope: CoroutineScope,
    ) {
        currentContentId = contentId
        currentTitle = title
        if (castSessionManager.sessionState.value.isConnected) {
            scope.launch {
                loadMediaToCast(streamUrl, contentType, subtitleTracks)
            }
        }
    }

    private fun handleStateChange(state: CastSessionState, scope: CoroutineScope) {
        when {
            state.isConnected -> {
                logger.info("Cast connected, starting sync")
                startPeriodicSync(scope)
            }
            else -> {
                syncJob?.cancel()
                syncJob = null
            }
        }
    }

    private fun startPeriodicSync(scope: CoroutineScope) {
        syncJob?.cancel()
        syncJob = scope.launch {
            while (isActive) {
                syncPlaybackState()
                delay(SYNC_INTERVAL_MS)
            }
        }
    }

    private suspend fun syncPlaybackState() {
        val player = attachedPlayer ?: return
        val positionMs = player.getCurrentPosition()
        val timeDiff = kotlin.math.abs(positionMs - lastSyncTimeMs)
        if (timeDiff < SYNC_THRESHOLD_MS) return

        val isPlaying = player.playerState.value is PlayerState.Playing
        val playbackState = CastPlaybackState(
            currentTime = positionMs,
            isPlaying = isPlaying,
            volume = 1.0f,
        )
        try {
            castSessionManager.syncPlaybackState(playbackState)
            lastSyncTimeMs = positionMs
        } catch (e: Exception) {
            logger.error("Failed to sync playback to cast", error = e)
        }
    }

    private suspend fun loadMediaToCast(
        streamUrl: String,
        contentType: String,
        subtitleTracks: List<CastSubtitleTrack>,
    ) {
        val media = CastMedia(
            contentId = currentContentId ?: streamUrl,
            title = currentTitle ?: "",
            streamUrl = streamUrl,
            posterUrl = null,
            duration = attachedPlayer?.getDuration(),
            subtitleTracks = subtitleTracks,
            contentType = contentType,
        )
        try {
            castSessionManager.loadMedia(media)
            logger.info("Media loaded to cast via bridge", mapOf("contentId" to media.contentId))
        } catch (e: Exception) {
            logger.error("Failed to load media to cast", error = e)
        }
    }

    companion object {
        private const val SYNC_INTERVAL_MS = 5000L
        private const val SYNC_THRESHOLD_MS = 1000L
    }
}
